imap/jmap_push_subscription.c

```diff
@@ -0,0 +1,762 @@
+    }
+
+    /* Serialize the subscription */
+    char *data = json_dumps(jpushsub, JSON_INDENT(2));
+    size_t datalen = strlen(data);
+
+    /* Create RFC 5322 header for subscription */
+    char *userid = mboxname_to_userid(mailbox_name(mailbox));
+    if (strchr(userid, '@')) {
+        buf_printf(&buf, "<%s>", userid);
+    }
+    else {
+        buf_printf(&buf, "<%s@%s>", userid, config_servername);
+    }
+    mimehdr = charset_encode_mimeheader(buf_cstring(&buf), buf_len(&buf), 0);
+    fprintf(f, "From: %s\r\n", mimehdr);
+    free(mimehdr);
+
+    mimehdr = charset_encode_mimeheader(id, 0, 0);
+    fprintf(f, "Subject: %s\r\n", mimehdr);
+    free(mimehdr);
+
+    time_to_rfc5322(now, datestr, sizeof(datestr));
+    fprintf(f, "Date: %s\r\n", datestr);
+
+    /* Use id@servername as Message-ID */
+    fprintf(f, "Message-ID: <%s@%s>\r\n", id, config_servername);
+
+    fputs("Content-Type: application/json; charset=utf-8\r\n", f);
+    fprintf(f, "Content-Length: " SIZE_T_FMT "\r\n", datalen);
+    fputs("Content-Disposition: attachment\r\n", f);
+    fputs("MIME-Version: 1.0\r\n", f);
+    fputs("\r\n", f);
+
+    /* Write the subscription data to the file */
+    fwrite(data, datalen, 1, f);
+    free(data);
+
+    fclose(f);
+
+    if (verified) {
+        strarray_append(&flags, "\\flagged");
+
+        /* Need authstate in order to set flags */
+        authstate = auth_newstate(userid);
+    }
+
+    if ((r = append_setup_mbox(&as, mailbox, userid, authstate,
+                               0, NULL, NULL, 0, EVENT_MESSAGE_NEW))) {
+        syslog(LOG_ERR, "append_setup(%s) failed: %s",
+               mailbox_name(mailbox), error_message(r));
+    }
+    else {
+        struct body *body = NULL;
+        struct timespec ts_expires = { expires, 0 };
+
+        /* Use internaldate to store expires time */
+        r = append_fromstage(&as, &body, stage, &ts_expires, 0, &flags, 0, NULL);
+        if (body) {
+            message_free_body(body);
+            free(body);
+        }
+
+        if (r) {
+            syslog(LOG_ERR, "append_fromstage() failed: %s", error_message(r));
+            append_abort(&as);
+        }
+        else {
+            /* Commit the append to the pushsub mailbox */
+            r = append_commit(&as);
+            if (r) {
+                syslog(LOG_ERR, "append_commit() failed: %s", error_message(r));
+            }
+        }
+    }
+
+    append_removestage(stage);
+    auth_freestate(authstate);
+    strarray_fini(&flags);
+    buf_free(&buf);
+    free(userid);
+
+    return r;
+}
+
+static strarray_t *allowed_push_types = NULL;
+
+static void validate_types(json_t *types, json_t *invalid)
+{
+    if (json_is_array(types)) {
+        size_t i;
+        json_t *val;
+
+        json_array_foreach(types, i, val) {
+            const char *type = json_string_value(val);
+
+            if (!type ||
+                strarray_find_case(allowed_push_types, type, 0) < 0) {
```

Comment:
In imapd.conf for `calendar_component_set` and `caldav_alarm_support_components` the value is again a space-separated list, but it seems the valid values are enforced by `BITFIELD`:

> { "calendar_component_set", "VEVENT VTODO VJOURNAL VFREEBUSY VAVAILABILITY VPOLL", BITFIELD("VEVENT", "VTODO", "VJOURNAL", "VFREEBUSY", "VAVAILABILITY", "VPOLL"), "3.1.7" }

BITFIELD might be reasonable also for imapoptions:jmap_push_types, possibly allowing to remove the `_case` here